Proof Refactoring

maul-esel opened the pull request:
This PR refactors how ULTIMATE works with proof artifacts (for now, primarily Floyd-Hoare annotations). The point is:
1. to make it easier to work with proofs, by making them first-class concepts (e.g. `IFloydHoareAnnotation`) rather than annotations in an Icfg. This allows e.g. considering the proofs of multiple CEGAR loops separately (a more elegant solution than the workaround used in #670).
2. to make the handling of proofs more flexible, in preparation for supporting other kinds of proofs; most importantly, _thread-modular proofs at many levels_ and _Owicki-Gries proofs_. These other kinds of proofs are e.g. needed for concurrency correctness witnesses.

I consider the PR more or less ready for review, if the currently running nightly tests don't show any more issues. The only change I still have planned from my side is to go through the `.epf` files and adapt them to the following change in the settings for `TraceAbstraction`:
> The setting `Compute Hoare Annotation of negated interpolant automaton, abstraction and CFG` has been removed. The existing setting `Positions where we compute the Hoare Annotation` has a possible value `NONE` (the default value) to disable Hoare computation completely.

However, I wanted to first ask if anyone objects to this or would propose a different settings structure, before I start going through all the settings files.

maul-esel committed:
begin sketching proof infrastructure: IProofProducer & friends

maul-esel committed:
IInitialAbstractionProvider should provide proof producers

maul-esel committed:
begin sketching proof production in CEGAR loops

maul-esel committed:
add operators for proof post processing and conversion

maul-esel committed:
start sketching Hoare proof producer

maul-esel committed:
cherry-pick Floyd-Hoare changes from wip/dk/empire-owicki

These changes allow computing Floyd/Hoare annotations for
product automata of concurrent programs.

maul-esel committed:
introduce IFloydHoareAnnotation

maul-esel committed:
move Floyd/Hoare handling

BasicCegarLoop should not refer to Floyd/Hoare proofs, let NwaCegarLoop
handle this. Also, CEGAR loops refer directly to Hoare annotation states,
not IcfgLocations.

maul-esel committed:
HoareAnnotationFragments: rely on supplied hoareAnnotationStates

Do not check the setting again, the caller should supply the appropriate states.

maul-esel committed:
move HoareAnnotationPositions to its own file

Also define their semantics in the class itself,
rather than in some utility class.

maul-esel committed:
begin defining interfaces for proof updates in NwaCegarLoop

maul-esel committed:
move NwaHoareProofProducer to its own file

maul-esel committed:
let CEGAR loops accept proof updaters

maul-esel committed:
add HoareAnnotationPositions.None

maul-esel committed:
specialize return type

maul-esel committed:
IFloydHoareAnnoation: decouple from automaton

maul-esel committed:
rename method

maul-esel committed:
move code for Floyd-Hoare annotations to package in Library-TraceCheckerUtils

maul-esel committed:
IFloydHoareAnnotation: remove unused type parameter

maul-esel committed:
FloydHoareMapping: remove unused automaton

maul-esel committed:
add Validity::and method

maul-esel committed:
add FloydHoareValidityCheck

maul-esel committed:
remove duplicate stopwatch stopping

maul-esel committed:
fix failures due to FloydHoare checking

maul-esel committed:
small simplifications in Hoare annotation classes

maul-esel committed:
continue refactoring from hard-coded Hoare annotations to proof producers

maul-esel committed:
simplify settings (will be modified further)

maul-esel committed:
add missing statistics result to initial abstraction providers

maul-esel committed:
missing license headers

maul-esel committed:
continue refactoring

maul-esel committed:
move proof-related code to its own library plugin

maul-esel committed:
fix NPE

maul-esel committed:
move more code to proofs plugin

maul-esel committed:
compute Hoare states from preferences

maul-esel committed:
some doc comments

maul-esel committed:
add new base class for proof producers

maul-esel committed:
better method name

maul-esel committed:
add check to AbstractProofProducer

maul-esel committed:
remove no longer needed dependency

maul-esel committed:
remove unused method

maul-esel committed:
remove now obsolete (and unused) parameter from CEGAR loops

maul-esel committed:
generalize IcfgAnnotationFromAutomaton to TransformFloydHoareAnnotation & document

maul-esel committed:
move printing of Floyd-Hoare annotations to a central location

maul-esel committed:
InvariantSynthesis: use FloydHoareMapping in place of HoareAnnotation

maul-esel committed:
deduplicate code for invariant and procedure contract results

maul-esel committed:
unify CodeCheck handling of invariants with TraceAbstraction, InvariantSynthesis

maul-esel committed:
minor simplifications

maul-esel committed:
remove unused param

maul-esel committed:
remove redundant constructor parameters

maul-esel committed:
small simplification

maul-esel committed:
some more small simplifications

maul-esel committed:
Merge branch 'dev' into wip/dk/proof-refactoring

# Conflicts:
#	trunk/source/InvariantSynthesis/src/de/uni_freiburg/informatik/ultimate/plugins/generator/invariantsynthesis/InvariantSynthesisStarter.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/smt/predicates/PredicateUtils.java
#	trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/floydhoare/HoareAnnotationComposer.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TraceAbstractionStarter.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TraceAbstractionUtils.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/predicates/HoareAnnotationChecker.java

maul-esel committed:
fix typo

maul-esel committed:
deprecate class

maul-esel committed:
remove HoareAnnotation class

maul-esel committed:
fix typo

maul-esel committed:
propagate proof updater down into AbstractCegarLoop

maul-esel committed:
rename and document parameter

maul-esel committed:
update some TODO comments

maul-esel committed:
narrow public signature of AbstractCegarLoop

maul-esel committed:
simplify type signatures

maul-esel committed:
separate proof production from postprocessing

CEGAR loops decide which kinds of proofs they provide and how.
Disentangle possible postprocessing of proofs from CEGAR loops
and proof production.

maul-esel committed:
work towards actually producing proofs

maul-esel committed:
Merge branch 'dev' into wip/dk/proof-refactoring

# Conflicts:
#	trunk/source/InvariantSynthesis/src/de/uni_freiburg/informatik/ultimate/plugins/generator/invariantsynthesis/InvariantSynthesisStarter.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/smt/predicates/PredicateUtils.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TraceAbstractionStarter.java

maul-esel committed:
Merge branch 'dev' into wip/dk/proof-refactoring

maul-esel committed:
deduplicate and unify helper methods

maul-esel committed:
fix copyright

maul-esel committed:
restore commented-out functionality

maul-esel committed:
cleanup and comments

maul-esel committed:
restore commented-out sanity check; delete obsolete interface

maul-esel committed:
continue refactoring; major simplifications

- introduce marker interface IProof to avoid using Object
- decouple AbstractCegarLoop, CegarLoopResult from proof production
- remove useless interfaces
- decouple proofs and proof translation from IInitialAbstractionProvider interface

maul-esel committed:
fix NPE and ConcurrentModificationException

maul-esel committed:
fix some NPEs: missing formulas

maul-esel committed:
fix concurrency automata setting: only apply for concurrent programs

maul-esel committed:
fix NPEs

maul-esel committed:
AbstractBuchiCegarLoop: precondition of interpolant automaton is not necessarily "true"

maul-esel committed:
fix settings: default to not producing Hoare annotation

maul-esel committed:
CodeCheckObserver: output proofs, like other plugins do

maul-esel committed:
use marker interface IProof

maul-esel committed:
some documentation and minor cleanups

maul-esel committed:
TraceAbstraction: don't forget to write proof into ICFG

maul-esel committed:
contract computation: requires may be needed even if ensures is null to prove asserts inside the procedure

maul-esel committed:
fix NPEs

maul-esel committed:
adapt setting files (epf, json) and CLI arguments to preferences change

bahnwaerter commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/IProof.java:
```diff
@@ -0,0 +1,36 @@
+ * @author Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ */
+public interface IProof {
+	// no methods here
```
Will proofs be able to implement interface functions in the future so that multiple implementations of a single proof is possible? If that is never the case, then an abstract class might be a more suitable generalization.

maul-esel replied:
Currently, there are no common functions planned for the immediate future. But in principle, some could be added -- e.g. a representation of the specification that is proved by the instance.

I do not see the point of an abstract class here. That would restrict implementations (they cannot inherit from other base classes), but what benefit would we get?

bahnwaerter replied:
Yeah your're right. I've seen later in the code review that my inital remark does not make sense anymore. Your model looks quite good to represent all possible objects related to proofs.

bahnwaerter commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/floydhoare/HoareAnnotationComposer.java:
```diff
@@ -284,8 +306,81 @@ public HoareAnnotationStatisticsGenerator getHoareAnnotationStatisticsGenerator(
+	/**
+	 * Construct Term which represents the same set of states as ps, but where all globalVars are renamed to
+	 * oldGlobalVars.
+	 *
```
Remove empty comment line!

bahnwaerter commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/floydhoare/HoareAnnotationPositions.java:
```diff
@@ -0,0 +1,70 @@
+			hoareAnnotationLocs.addAll(IcfgUtils.getReturnPredecessorPoints(root));
+			break;
+		case None:
+			return Set.of();
```
You can also return an empty set using `Collections.emptySet()` if it should be immutable.

maul-esel replied:
I've wondered about this in other places before -- do you know if either `Set.of()`  or `Collections.emptySet()` has an advantage over the other?

bahnwaerter replied:
To my knowledge, there is a slight difference in the implementation:

````
Collections.emptySet().contains(null); // false
Set.of().contains(null); // NullPointerException
````

Also the serialized form might be different.

bahnwaerter commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/AbstractCegarLoop.java:
```diff
@@ -512,6 +488,10 @@ private void refineAbstractionInternal(final AutomatonType automatonType)
 		}
 	}
 
+	protected void performAbstractionSanityCheck() {
+		// Empty implementation. Subclasses may override this method.
+	}
```
Maybe declare it as an abstract method and add implementation in the subclasses?

maul-esel replied:
That would force every subclass to provide an implementation. Most of them currently do not have anything meaningful to do here.

bahnwaerter replied:
Okay, I see. Your implemented version would be the better choice.

bahnwaerter commented on trunk/source/TraceAbstractionConcurrent/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstractionconcurrent/TraceAbstractionConcurrentObserver.java:
```diff
@@ -142,15 +142,15 @@ public boolean process(final IElement root) {
 		clrReporter.reportAllSafeResultIfNecessary(result, errNodesOfAllProc.size());
 
-		mLogger.info("Statistics - iterations: " + abstractCegarLoop.getIteration());
+		// mLogger.info("Statistics - iterations: " + abstractCegarLoop.getIteration());
```
Shouldn't be that printed anymore? If so, you can remove this line.

maul-esel replied:
Sure, removed. We should probably get around to deleting the entire project soon, as it is obsolete and unused, but complicates implementation of other projects.

bahnwaerter commented on trunk/source/TraceAbstractionConcurrent/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstractionconcurrent/TraceAbstractionConcurrentObserver.java:
```diff
@@ -142,15 +142,15 @@ public boolean process(final IElement root) {
 		String stat = "";
 		stat += "Statistics:  ";
-		stat += " Iterations " + abstractCegarLoop.getIteration() + ".";
+		// stat += " Iterations " + abstractCegarLoop.getIteration() + ".";
```
Shouldn't be that printed anymore? If so, you can remove this line.

bahnwaerter reviewed (approved):
The introduction of the new library for proofs looks great, at least for my knowledge on the subject. The new proof library has been created correctly. We just need to upgrade the Java version number in the `Manifest.MF` file later when we finally migrate to Java 21. In addition, @danieldietsch has to adjust the sonar coverage configuration in the Maven `pom.xml` under `trunk/source/BA_MavenParentUltimate` so that the code coverage for the library can be reported properly.

maul-esel committed:
minor cleanup

maul-esel committed:
Merge branch 'dev' into wip/dk/proof-refactoring

Removes the workaround introduced in 720a012, as proofs of different
CEGAR loops are now stored separately in the ICFG.

# Conflicts:
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/NwaCegarLoop.java

maul-esel commented:
Thanks for the reviews. I'll wait for the nightly test run; if no further issues occur I will merge this tomorrow.

maul-esel committed:
FloydHoareValidityCheck: add more informative failure messages

maul-esel committed:
remove unnecessary type parameter from results

maul-esel committed:
add explicit representations of the specification shown by a proof

This is particularly useful if we create multiple proof objects
for the same program and different (partial) specifications, to
distinguish which proof objects relate to which specification.

maul-esel committed:
PrePostConditionSpecification: precondition may depend on initial location

maul-esel committed:
include Check objects in InvariantResult / ProcedureContractResult

maul-esel committed:
fix block encoding bug discovered by contract computation

Every procedure that exists in the IIcfg should have an exit location,
even if that location is unreachable.

This was not the case for the IIcfgs produced by the BlockEncodingV2 plugin.
This commit fixes this bug, and introduces an assert statement after block
encoding has finished to check if the produced IIcfg satisfies this invariant.

Moreover, we add two sanity checks regarding the computation of procedure contracts,
where the problem was discovered (the location of the ProcedureContractResult was null),
such that similar issues are caught earlier in the future.

maul-esel commented:
Fixing the tests required some changes:
- The precondition for our Floyd-Hoare annotations is not necessarily `true`; instead, it is a conjunction of `(= g |old(g)|)` for all modifiable global variables `g` of the initial location's procedure. This requires supporting different preconditions per initial location, in the case of multiple initial locations.
- A bug in `BlockEncodingV2` caused contract computation to fail, as some procedures did not have an exit location. I've fixed this, and added some checks to prevent similar bugs in the future.

While doing this, I also added a feature required e.g. in the context of requirements analysis: Each proof now also has an explicit description of the specification that it proves. This information is also added to `InvariantResult` and `ProcedureContractResult` (in the form of `Check` objects).

danieldietsch commented on releaseScripts/default/adds/Ultimate.py:
```diff
@@ -508,9 +508,9 @@ def create_cli_settings(prop, validate_witness, architecture, c_file):
         # but only for reachability witness validation
         ret.append(
-            "--traceabstraction.compute.hoare.annotation.of.negated.interpolant.automaton,.abstraction.and.cfg"
+            "--traceabstraction.positions.where.we.compute.the.hoare.annotation"
```
To be honest, I dont like options that huamnize Ultimate, i.e., "where we compute".

maul-esel replied:
:+1: I fully agree. This is an existing option, so for this PR I would keep it. When we do the settings restructuring that we discussed, we should change it.

danieldietsch commented on trunk/examples/LTL/regression/c/LTLAutomizerC.epf:
```diff
@@ -205,7 +205,7 @@ file_export_version=3.0
 /instance/de.uni_freiburg.informatik.ultimate.plugins.blockencoding/Create\ interprocedural\ compositions=false
 /instance/de.uni_freiburg.informatik.ultimate.plugins.blockencoding/Iterate\ optimizations\ for\ n\ times\ (<\=0\ means\ until\ nothing\ changes)=0
 #Fri Dec 09 13:35:39 CET 2016
-/instance/de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction/Positions\ where\ we\ compute\ the\ Hoare\ Annotation=All
```
Why remove the explicit setting here?

maul-esel replied:
The setting is not removed, the value is changed to `None`. The reason is that the total effect of this setting with the removed line `/instance/de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction/Compute\ Hoare\ Annotation\ of\ negated\ interpolant\ automaton,\ abstraction\ and\ CFG=false` below was to disable Hoare computation completely, which is now simply achieved with the setting value `None`.

danieldietsch commented on trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/cegar/BuchiCegarLoopFactory.java:
```diff
@@ -116,8 +117,8 @@ public BuchiCegarLoopFactory(final IUltimateServiceProvider services, final TAPr
-			final var automatonProvider = new Petri2FiniteAutomatonAbstractionProvider.Lazy<>(petriNetProvider,
-					stateFactoryForRefinement, new AutomataLibraryServices(mServices));
+			final var automatonProvider = new Petri2FiniteAutomatonAbstractionProvider.Lazy<>(mServices,
+					petriNetProvider, stateFactoryForRefinement);
```
👍

danieldietsch commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/floydhoare/FloydHoareMapping.java:
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2023 Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
```
Sitting there for a while :)

danieldietsch commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/PrePostConditionSpecification.java:
```diff
@@ -0,0 +1,117 @@
+
+/**
+ * A specification stating that any execution starting in an initial location and with a variable assignment satisfying
+ * a given precondition must, when (and if) it reaches a final location, have a current variable assignment satisfying a
```
What is it, when or if?

maul-esel replied:
I've reworded the text to be hopefully more clear.

danieldietsch commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/PrePostConditionSpecification.java:
```diff
@@ -0,0 +1,117 @@
+public class PrePostConditionSpecification<S> implements ISpecification {
+	private final Map<S, IPredicate> mInitialStates;
+	private final Predicate<S> mIsFinalState;
+	private final IPredicate mPostcondition;
```
* The initial locations are explicitly specified, but the final locations use a predicate. Is there a reason for that? Might be worth a comment :). 
* Are there any constraints on the number of final locations? Are these final locations in the automata-sense, or are these exit locations of the program?
* Am I understanding the idea here correctly: for each initial state and its precondition, all reachable final states satsify the post condition? So, if I wanted to split a program (i.e., construct a residual program), I would construct two of these specifications, one for the postcondition `false` and all initial states that guarantee that, and one for `true`. And I would then think that "final states" are not the exits of the program (which would be somewhat natural when talking about pre/postconditions), but the accepting states of the automaton, i.e., our error locations.

maul-esel replied:
I've added some comments to hopefully clarify the first two questions.

I don't fully understand the 3rd question: Are you saying you would create a specification with _postcondition_ `true` ? That specification would be trivial, so it's perhaps not what you intended.
- I could split the specification by splitting the set of initial states, but keeping the same postcondition.
- Alternatively, I could split it by splitting the set of final states, again with the same postcondition.

maul-esel committed:
Merge branch 'dev' into wip/dk/proof-refactoring

# Conflicts:
#	trunk/source/InvariantSynthesis/src/de/uni_freiburg/informatik/ultimate/plugins/generator/invariantsynthesis/InvariantSynthesisStarter.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/smt/predicates/HoareAnnotation.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/smt/predicates/PredicateFactory.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/HoareAnnotationWriter.java

danieldietsch commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/floydhoare/FloydHoareMapping.java:
```diff
@@ -0,0 +1,79 @@
+	private final Map<S, IPredicate> mAnnotation;
+	private final IPredicate mDefaultPredicate;
+
+	public FloydHoareMapping(final PrePostConditionSpecification<S> specification,
```
Is there a reason to support both, null and a default predicate? Wouldnt it be easier to force a default predicate?

maul-esel replied:
We use `null` to indicate that a location does not have any annotation. Since we have settings to only compute the annotation at certain locations (e.g. loop heads), this can be the case for ICFGs; and there is no non-null default predicate we could use in this case.

danieldietsch replied:
Its just that having no annotation would mean I could always use true in that case, right?

maul-esel replied:
No, that would be different. Imagine an edge `src --(x:=0)--> tgt`, where `tgt` has an annotation like `x==y`. If we check the validity of this annotation (as in `FloydHoareValidityChecker`), we have to differentiate:
- If `src` has the annotation `true` (perhaps as default predicate), the Hoare triple `{true} x:=0 {x==y}` is not valid, and so the entire annotation is invalid.
- If `src` has the annotation `null`, it is clear that we do not have the full Hoare triple to check. But we don't consider the annotation invalid in this case.

danieldietsch replied:
Ah, perhaps this should be a comment for this class :)

maul-esel replied:
:+1: I've added this explanation to `IFloydHoareAnnotation` and linked to it in the documentation for the two different constructors here.

danieldietsch reviewed (commented):
Review so far, I would read on, but feel free to merge anyways,

maul-esel committed:
extended documentation of PrePostConditionSpecification

danieldietsch commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/PrePostConditionSpecification.java:
```diff
@@ -42,23 +42,40 @@
+ * error locations of a program, or the accepting states of an interpolant automaton, together with the postcondition
+ * {@code false}. In this case, the specification expresses unreachability of these states (see
+ * {@link #isUnreachabilitySpecification()}). Alternatively, one might use the exit location of a program (or procedure)
+ * with a postcondition specifying the expected variable values when the program terminates.
```
👍

danieldietsch commented on trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/PrePostConditionSpecification.java:
```diff
@@ -42,23 +42,40 @@
+	//
+	// These are not listed explicitly but only identified by a boolean function, because in some cases (e.g. concurrent
+	// programs) the set of final states can be very large and costly to compute up-front.
 	private final Predicate<S> mIsFinalState;
```
:+1:

maul-esel committed:
more documentation

maul-esel committed:
Merge branch 'dev' into wip/dk/proof-refactoring

# Conflicts:
#	releaseScripts/website-config/frontend/config.js
#	trunk/source/WebsiteStatic/config/config.dist.js

maul-esel committed:
fix Hoare settings in new settings files
